### PR TITLE
pnp: move shadow loggers to parent module

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92058
* #91639
* __->__ #91428
* #91421

Summary:

Before this PR, PNP added shadow loggers to insides of
the shadow wrapper modules.

This PR moves those loggers to the parent module.

There are a couple of benefits:
1. this will unbreak features of quantization API which don't support loggers (such as hardcoding model output to be quantized)
2. this makes it easier to look at the parent graph and visualize what is logged, since now all the logging is in the same graph
3. this will make it easier to implement features such as propagation error calculation in the future

Test plan:

```
python test/test_quantization.py -k NShadows
```

### CODE REVIEW

File: torch/ao/ns/fx/n_shadows_utils.py [hunk 1]
```diff
@@ -593,16 +523,14 @@ def handle_subgraph_candidate(
...
             # TODO(future PR): handle fusion patterns where non-first nodes
             # need inputs
 
             # pass in all node args and kwargs
 
-            # the first argument is always the reference output of the last
-            # node of this subgraph
-            new_args = [last_node]
-
+            new_args = []
```
Comment (jerryzh168) [inline torch/ao/ns/fx/n_shadows_utils.py:533]: just for my understanding, why is the `last_node` argument removed?

**Review by jerryzh168 — APPROVED** [review]
looks good, had some inline comments

File: torch/ao/ns/fx/n_shadows_utils.py (hunk 1 above)
Comment (vkuzo) [reply torch/ao/ns/fx/n_shadows_utils.py:533]: Before it was

```
# parent
x0 -> op -> x1 
  \                    \
    -----------shadow

# shadow
x0 -> op_shadow -> x1_shadow -> log 
                                                            /
x1 --------------------------------
```

Now, it's

```
# parent
x0 -> op -> x1 ---------> log
  \                                         /  
    -shadow -> x1_shadow


# shadow
x0 -> op_shadow -> x1_shadow
```

the logger is moved outside, and `last_node` is the thing which the logger needs to compare actual to reference

File: torch/ao/ns/fx/n_shadows_utils.py [hunk 2]
```diff
@@ -705,11 +647,21 @@ def handle_subgraph(
 
     fqn = _maybe_get_fqn(first_node, mt)
 
+    # We want the results to contain the subgraphs in natural order,
```
[the remark this answers is not in the document]
Comment (vkuzo) [reply torch/ao/ns/fx/n_shadows_utils.py:650]: just tried it, unfortunately it leads to the same issue explained below in the comment :(

**Pull request closed without merging**